flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/HsSelectiveSpillingStrategyTest.java

```diff
@@ -0,0 +1,163 @@
+
+    public static final float SELECTIVE_SPILL_BUFFER_RATIO = 0.4f;
+
+    @BeforeEach
+    void before() {
+        this.spillStrategy =
+                new HsSelectiveSpillingStrategy(
+                        HybridShuffleConfiguration.builder(NUM_SUBPARTITIONS, 1)
+                                .setSelectiveStrategySpillThreshold(SELECTIVE_SPILL_THRESHOLD)
+                                .setSelectiveStrategySpillBufferRatio(SELECTIVE_SPILL_BUFFER_RATIO)
+                                .build());
+    }
+
+    @Test
+    void testOnBufferFinished() {
+        BufferWithIdentity bufferWithIdentity = new BufferWithIdentity(createBuffer(), 0, 0);
+        Action finishedAction =
+                spillStrategy.onBufferFinished(spillInfoProvider, bufferWithIdentity);
+        assertThat(finishedAction.isImmediate()).isTrue();
+        assertThat(finishedAction.getBufferToRelease()).isEmpty();
+        assertThat(finishedAction.getBufferToSpill()).isEmpty();
+    }
+
+    @Test
+    void testOnBufferConsumed() {
+        BufferWithIdentity bufferWithIdentity = new BufferWithIdentity(createBuffer(), 0, 0);
+        Action consumedAction =
+                spillStrategy.onBufferConsumed(spillInfoProvider, bufferWithIdentity);
+        assertThat(consumedAction.isImmediate()).isTrue();
+        assertThat(consumedAction.getBufferToRelease()).hasSize(1);
+        assertThat(consumedAction.getBufferToRelease()).element(0).isEqualTo(bufferWithIdentity);
+        assertThat(consumedAction.getBufferToSpill()).isEmpty();
+    }
+
+    @Test
+    void testOnUsedMemoryChanged() {
+        testOnUsedMemoryLow();
+        testOnUsedMemoryHigh();
+    }
+
+    void testOnUsedMemoryLow() {
+        final int bufferPoolSize = 10;
+        final int bufferThreshold = (int) (bufferPoolSize * SELECTIVE_SPILL_THRESHOLD);
+        spillInfoProvider.setCurrentPoolSizeSupplier(() -> bufferPoolSize);
+        spillInfoProvider.setNumTotalUnSpilledBuffersSupplier(() -> bufferThreshold - 1);
+        Action action = spillStrategy.onUsedMemoryChanged(spillInfoProvider);
+        assertThat(action.isImmediate()).isTrue();
+        assertThat(action.getBufferToSpill()).isEmpty();
+        assertThat(action.getBufferToRelease()).isEmpty();
+    }
+
+    void testOnUsedMemoryHigh() {
+        final int subpartition1 = 0;
+        final int subpartition2 = 1;
+        final int subpartition3 = 2;
+
+        Map<Integer, Deque<BufferWithIdentity>> subpartitionBuffers = new HashMap<>();
+        subpartitionBuffers.put(subpartition1, createBufferWithIdentities(subpartition1, 0, 4, 6));
+        subpartitionBuffers.put(subpartition2, createBufferWithIdentities(subpartition2, 3, 7, 9));
+        subpartitionBuffers.put(subpartition3, createBufferWithIdentities(subpartition3, 5, 6, 11));
+
+        spillInfoProvider.setUnSpilledBuffersFunction(subpartitionBuffers::get);
+        spillInfoProvider.setConsumptionProgressSupplier(() -> Arrays.asList(-1, 1, 2));
```

Comment:
It's not easy to tell which buffer is expected to be spilled with higher priority.

I'd suggest to set the consuming progress with constants:
```
final int progress1 = 10;
final int progress2 = 20;
final int progress3 = 30;
```

And set the buffer index relative to the consuming progress:
```
createBufferWithIdentities(subpartition1, progress1 + 0, progress1 + 3, progress1 + 6);
createBufferWithIdentities(subpartition2, progress2 + 1, progress2 + 4, progress2 + 7);
createBufferWithIdentities(subpartition3, progress3 + 2, progress3 + 5, progress3 + 8);
```

Reply:
Then you won't need `getBufferAndScores` and can directly tell which buffers are expected to be spilled.

Reply:
Good idea! Changed to this way.